src/self_driving/modeling/operating_unit_recorder.cpp

```diff
@@ -49,10 +49,69 @@
 #include "self_driving/modeling/operating_unit.h"
 #include "self_driving/modeling/operating_unit_util.h"
 #include "storage/block_layout.h"
+#include "storage/index/index.h"
 #include "type/type_id.h"
 
+/**
+ * Used for grouping together indexes during IndexInsert/IndexDelete
+ * based on the (index size, key_size, num_keys).
+ */
+using index_op_key_t = std::tuple<size_t, size_t, size_t>;
+
+namespace std {
+template <>
+struct hash<index_op_key_t> {
+  std::size_t operator()(index_op_key_t const &key) const noexcept {
+    noisepage::common::hash_t hash = noisepage::common::HashUtil::Hash(std::get<0>(key));
+    hash = noisepage::common::HashUtil::CombineHashes(hash, noisepage::common::HashUtil::Hash(std::get<1>(key)));
+    hash = noisepage::common::HashUtil::CombineHashes(hash, noisepage::common::HashUtil::Hash(std::get<2>(key)));
+    return hash;
+  }
+};
+}  // namespace std
+
 namespace noisepage::selfdriving {
 
+template <typename IndexPlanNode>
+void OperatingUnitRecorder::RecordIndexOperations(const std::vector<catalog::index_oid_t> &index_oids) {
+  selfdriving::ExecutionOperatingUnitType type;
+  if (std::is_same<IndexPlanNode, planner::InsertPlanNode>::value) {
+    type = selfdriving::ExecutionOperatingUnitType::INDEX_INSERT;
+  } else if (std::is_same<IndexPlanNode, planner::DeletePlanNode>::value) {
+    type = selfdriving::ExecutionOperatingUnitType::INDEX_DELETE;
+  } else if (std::is_same<IndexPlanNode, planner::UpdatePlanNode>::value) {
+    // UPDATE is done as a DELETE followed by INSERT
+    RecordIndexOperations<planner::InsertPlanNode>(index_oids);
+    RecordIndexOperations<planner::DeletePlanNode>(index_oids);
+    return;
+  } else {
+    NOISEPAGE_ASSERT(false, "Recording index operations for non-modiying plan node");
+  }
+
+  // Group indexes together
+  std::unordered_map<index_op_key_t, size_t> tracking;
+  for (auto &oid : index_oids) {
+    auto &index_schema = accessor_->GetIndexSchema(oid);
+
+    // Currenty favor using the index size extracted from the index as opposed to table size
+    // estimate. This will probably be replaced with stats in the future.
+    auto index = accessor_->GetIndex(oid);
+
+    std::vector<catalog::indexkeycol_oid_t> keys;
+    size_t num_keys = index_schema.GetColumns().size();
+    size_t key_size = ComputeKeySize(common::ManagedPointer(&index_schema), false, keys, &num_keys);
+    tracking[std::make_tuple(index->GetSize(), key_size, num_keys)] += 1;
+  }
+
+  // Record each distinct group
+  for (auto &track : tracking) {
+    // TODO(wz2): Eventually need to modify track.second to consider num_rows/cardinality
+    pipeline_features_.emplace(
+        type, ExecutionOperatingUnitFeature(current_translator_->GetTranslatorId(), type, std::get<0>(track.first),
+                                            std::get<1>(track.first), std::get<2>(track.first), track.second, 1, 0, 0));
+  }
```

Comment:
I actually don't think we need to group the indexes here. Typically a table won't have a hundred different indexes. I think it's simpler to just leave the indexes separately.